aws/hll/aws-s3-transfer-manager/src/download/discovery.rs

```diff
@@ -0,0 +1,328 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::ops::RangeInclusive;
+use std::str::FromStr;
+use std::{cmp, mem};
+
+use aws_sdk_s3::operation::get_object::builders::GetObjectInputBuilder;
+use aws_smithy_types::body::SdkBody;
+use aws_smithy_types::byte_stream::{AggregatedBytes, ByteStream};
```

Comment:
I think these are re-exported from `aws_sdk_s3::primitives`, so we should prefer using them wherever possible. That way, we don't need to depend on a core smithy runtime crate to get them.

Reply:
Good question, I'd say it's preferable, as mentioned in [README](https://github.com/smithy-lang/smithy-rs/tree/50c825b5dbd3471bc731d7a8fbeaae78fe5a2321/rust-runtime/aws-smithy-types). Because an SDK crate is usually the one taking direct dependency on runtime crates, whoever depending on the SDK crate shouldn't need to depend on them directly (that's sort of bypassing dependency laying). John was a proponent to use re-exports as much as possible, and we've had issues/PRs to clean that up (like [this](https://github.com/smithy-lang/smithy-rs/pull/3105) or [a list of those](https://github.com/search?q=repo%3Asmithy-lang%2Fsmithy-rs+export&type=pullrequests)).

In this case, the types in question are already re-exported, so might get them from `aws_smithy_types` and remove a direct dependency on `aws_smithy_types`, if not used after cleanup.

Reply:
Good question, I'd say it's preferable, as mentioned in [README](https://github.com/smithy-lang/smithy-rs/tree/50c825b5dbd3471bc731d7a8fbeaae78fe5a2321/rust-runtime/aws-smithy-types). Because an SDK crate is usually the one taking direct dependency on runtime crates, whoever depending on the SDK crate shouldn't need to depend on them directly (that's sort of bypassing dependency layering). John was a proponent to use re-exports as much as possible, and we've had issues/PRs to clean that up (like [this](https://github.com/smithy-lang/smithy-rs/pull/3105) or [a list of those](https://github.com/search?q=repo%3Asmithy-lang%2Fsmithy-rs+export&type=pullrequests)).

In this case, the types in question are already re-exported, so might get them from `aws_smithy_types` and remove a direct dependency on `aws_smithy_types`, if not used after cleanup.

Reply:
Ok I'll add a TODO for us to reconsider this. I don't (currently anyway) see the harm here as we're building up a new library as an abstraction on lower level parts (one of which is the `aws_sdk_s3` crate). I don't personally see any harm at the moment in re-using public types from the underlying runtime crates (especially types we _expect_ users to consume) but I also don't have a compelling reason to not follow your advice here.